cores/esp32/MacAddress.cpp

```diff
@@ -0,0 +1,230 @@
+    _mac.bytes[2] = b3;
+    _mac.bytes[3] = b4;
+    _mac.bytes[4] = b5;
+    _mac.bytes[5] = b6;
+}
+
+MacAddress::MacAddress(uint8_t b1, uint8_t b2, uint8_t b3, uint8_t b4, uint8_t b5, uint8_t b6, uint8_t b7, uint8_t b8) {
+    _type = MAC8;
+    _mac.bytes[0] = b1;
+    _mac.bytes[1] = b2;
+    _mac.bytes[2] = b3;
+    _mac.bytes[3] = b4;
+    _mac.bytes[4] = b5;
+    _mac.bytes[5] = b6;
+    _mac.bytes[6] = b7;
+    _mac.bytes[7] = b8;
+}
+
+//Parse user entered string into MAC address
+bool MacAddress::fromString(const char *buf) {
+    if(strlen(buf) == 17) {
+        return fromString6(buf);
+    } else if(strlen(buf) == 23) {
+        return fromString8(buf);
+    }
+    return false;
+}
+
+//Parse user entered string into MAC address
+bool MacAddress::fromString6(const char *buf) {
+    char cs[18];
+    char *token;
+    char *next; //Unused but required
+    int i;
+
+    strncpy(cs, buf, sizeof(cs)); //strtok modifies the buffer: copy to working buffer.
+
+    for(i = 0; i < 6; i++) {
+        token = strtok((i==0) ? cs : NULL, ":");    //Find first or next token
+        if(!token) {                                //No more tokens found
+            return false;
+        }
+        _mac.bytes[i] = strtol(token, &next, 16);
+    }
+    _type = MAC6;
+    return true;
+}
+
+bool MacAddress::fromString8(const char *buf) {
+    char cs[24];
+    char *token;
+    char *next; //Unused but required
+    int i;
+
+    strncpy(cs, buf, sizeof(cs)); //strtok modifies the buffer: copy to working buffer.
+
+    for(i = 0; i < 8; i++) {
+        token = strtok((i==0) ? cs : NULL, ":");    //Find first or next token
+        if(!token) {                                //No more tokens found
+            return false;
+        }
+        _mac.bytes[i] = strtol(token, &next, 16);
+    }
+    _type = MAC8;
+    return true;
+}
+
+//Copy MAC into byte array
+void MacAddress::toBytes(uint8_t *buf) {
+    if(_type == MAC6) {
+        memcpy(buf, _mac.bytes, 6);
+    } else {
+        memcpy(buf, _mac.bytes, sizeof(_mac.bytes));
+    }
+}
+
+//Print MAC address into a C string.
+//MAC: Buffer must be at least 18 chars
+int MacAddress::toString(char *buf) {
+    if(_type == MAC6) {
+        return sprintf(buf, "%02X:%02X:%02X:%02X:%02X:%02X",
+            _mac.bytes[0], _mac.bytes[1], _mac.bytes[2],
+            _mac.bytes[3], _mac.bytes[4], _mac.bytes[5]);
+    } else {
+        return sprintf(buf, "%02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X",
+            _mac.bytes[0], _mac.bytes[1], _mac.bytes[2],
+            _mac.bytes[3], _mac.bytes[4], _mac.bytes[5],
+            _mac.bytes[6], _mac.bytes[7]);
+    }
+}
+
+String MacAddress::toString() const {
+    if(_type == MAC6) {
+        char buf[18];
+        sprintf(buf, "%02X:%02X:%02X:%02X:%02X:%02X",
+            _mac.bytes[0], _mac.bytes[1], _mac.bytes[2],
+            _mac.bytes[3], _mac.bytes[4], _mac.bytes[5]);
+        return String(buf);
+    } else {
+        char buf[24];
```

Comment:
use one buf and one return, just change the sprintf in the if (and actually use `snprintf` to limit the size)

Reply:
done